### DISCUSSION TITLE
Code execution agent returning empty message and being called recursively

### DISCUSSION THREAD

**Discussion by gvijqb** [microsoft/autogen#4797] [Q&A]
Hey guys,

I have a code execution agent defined with user proxy agent as follows:

```python
self.executor = autogen.UserProxyAgent(name="Executor",system_message=executor_msg,human_input_mode="NEVER",code_execution_config={"last_n_messages": 2,"executor": self.executor},)
```

Sometimes, the executor agent just ends up repeatedly being called by speaker and executor keeps repeating empty message as shown below:
<img width="1013" alt="Screenshot 2024-12-23 at 11 38 27 PM" src="https://github.com/user-attachments/assets/ea2346dc-1110-4160-bde5-634ccfd68a91" />

Autogen version:
autogen-agentchat                        0.2.39

What can I do about this?

**Comment by ekzhu** [comment]
You may want to check the Python code it is running. 

You can set `allow_repeat_speaker =False` in group chat to disable the same speaker getting selected.

**Reply by gvijqb** [reply]
Sure, based on your suggestion I tried the `allow_repeat_speaker=False` but still it is going into a repetitive loop of empty executor messages. 

I observed that during this phase, no python code was being executed by it. Instead it just keeps sending a POST request to openAI chat completion as can be seen in the screenshot.

```
INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions "HTTP/1.1 200 OK"
```

Have you ever noticed this behavior?

**Reply by ekzhu** [reply]
Likely your user proxy is making LLM calls instead of executing code since no code was found in the message? 

Though LLM call can also happen when the group chat manager selects next speaker

**Reply by gvijqb** [reply]
> Likely your user proxy is making LLM calls instead of executing code since no code was found in the message?
> 
> Though LLM call can also happen when the group chat manager selects next speaker

Yes the behaviour appears to be random. If there was no code present in the message then the speaker selector should not call executor. Not sure how to debug this. 

Also, if you notice, the empty messages started appearing when the above agent tried calling a function which doesn't exist. Then this repeatedly kept calling executor. I am not sure why it called a tool which doesn't exist in the first place.

**Reply by ekzhu** [reply]
LLM can be random, and the GroupChatManager is using an LLM to make the choice.

v0.4 is set to come out of dev in January, and it restricts tool calling within individual agent, it should help to address this issue.

**Reply by gvijqb** [reply]
> LLM can be random, and the GroupChatManager is using an LLM to make the choice.
> 
> v0.4 is set to come out of dev in January, and it restricts tool calling within individual agent, it should help to address this issue.

Definitely looking forward to 0.4 release in that case.
